File: 2014/Round_1B/Repeater/repeater.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

string shortify(vector<char> data) {
  vector<char> shorty;
  shorty.push_back(data[0]);
  for (int i = 0; i < data.size(); ++i) {
     if (data[i] != shorty[shorty.size()-1]) {
       shorty.push_back(data[i]);
     }
   }
  string shortstr = "";
  for (int i = 0; i < shorty.size(); ++i) {
    shortstr+=shorty[i];
  }
  return shortstr;
}

int difference(vector<int> data) {
  int count = 0, N = data.size();
  sort(data.begin(), data.end());
  for (int i = 0; i < data.size(); ++i) {
    if (data[i] < data[N/2]) count+=(data[N/2]-data[i]);
    if (data[i] > data[N/2]) count+=(data[i]-data[N/2]);
  }
  return count;
}

int average(vector< vector<int> > data) {
  int count = 0;
    for (int i = 0; i < data[0].size(); ++i) {
      vector<int> tmp;
      for (int j = 0; j < data.size(); ++j)
        tmp.push_back(data[j][i]);
      count+=difference(tmp);
    }
  return count;
}

int solve(int n, vector< vector<char> > data) {
  string shortver = shortify(data[0]);
  bool win = true;
  vector< vector<int> > result;
  for (int i = 0; i < data.size(); ++i) {
    vector<int> totalcounts;
    for (int letter = 0; letter < shortver.length(); ++letter) {
      int count = -1;
      while (data[i].size() > 0) {
        if (data[i][0] == shortver[letter]) {
          data[i].erase(data[i].begin());
          count+=1;
        } else { break; }
      }
      if ( count == -1 ) {
        win = false;
        break;
      } else { totalcounts.push_back(count); }
    }
    if (data[i].size() >= 1) { win = false; }
    if (!win) {
      break;
    } else {
      result.push_back(totalcounts);
    }
  }
  if (!win) return -1;
  return average(result);
}

int main(int argc, char const * argv[]) {
  freopen("input.in", "r", stdin);
  int test;
  cin >> test;
  for (int t = 1; t <= test; ++t) {
    int n;
    cin >> n;
    vector< vector<char> > str;
    for (int i = 0; i < n; ++i) {
      string tp;
      cin >> tp;
      vector<char> temp;
      for (int j = 0; j < tp.length(); ++j)
        temp.push_back(tp[j]);
      str.push_back(temp);
    }
    int res = solve(n, str);
    if (res == -1) {
      cout << "Case #" << t << ": Fegla Won" << endl;
    } else { cout << "Case #" << t << ": " << res << endl; }
  }
  return 0;
}

File: 2014/Round_1B/Repeater/repeater.py
#!/usr/bin/env python
# -*- coding: utf8 -*-

inputfile = 'input.in'
outputfile = 'output.op'

def shortify(data):
  shortver = data[0][0]
  for i in data[0]:
    if i != shortver[-1]:
      shortver+=i
  return shortver

def average(data):
  count=0
  for i in xrange(len(data[0])):
    tmp = []
    for j in xrange(len(data)):
      tmp.append(data[j][i])
    count+=difference(tmp)
  return count

def difference(lst):
  lst = map(int, sorted(lst))
  count, N = 0, len(lst)-1
  for i in lst:
    if i < lst[N/2]: count+=(lst[N/2]-i)
    elif i > lst[N/2]: count+=(i-lst[N/2])
  return count

def solve(n, data):
  shortver = shortify(data)
  win = True
  result = []
  for i in xrange(len(data)):
    totcounts = []
    for letter in shortver:
      count = -1
      while len(data[i]) > 0:
        if data[i][0] == letter:
          data[i]=data[i][1:]
          count+=1
        else: break
      if count == -1:
        win = False
        break
      else:totcounts.append(count)
    if len(data[i]) >= 1: win = False
    if not win: break
    else: result.append(totcounts)
  if not win: return 'Fegla Won'
  return str(average(result))

if __name__ == '__main__':
  inputdata = open(inputfile, 'rb').readlines()
  cases = int(inputdata[0])
  inputdata.pop(0)
  with open(outputfile, 'wb') as out:
    for case in range(cases):
      n = int(inputdata[0])
      data = [x.strip() for x in inputdata[1:n+1]]
      out.write('Case #%d: %s\n' % (case+1, solve(n, data)))
      inputdata = inputdata[n+1:]